### PR TITLE
Change phase duration to maximum phase duration

### PR DESCRIPTION
**Pull request by eightarmsninebrains**
In the previous version, there was a single parameter `phase_duration_name` for each name. The model assumed that all such parameters were fixed values.  

This change:
- creates  parameters `phase_duration_name_min_blocks` and `phase_duration_name_max_blocks` for each name. 
- uses "phase_duration_name_max_blocks" to determine if the phase duration has been exceeded. 

The next steps are:
- use `phase_duration_name_min_blocks` in the logic. (Make sure that the logic has the intended effect.)
- create methods that make it easier to construct parameters with specific values, for parameter sweeps and experiments.

### CODE REVIEW

**Comment by eightarmsninebrains** [comment]
Further updates to create time series signals for shock scenarios, and make them compatible with the current `GasEstimator` architecture.

File: aztec_gddt/logic.py [hunk 1]
```diff
@@ -372,25 +375,29 @@ def p_commit_bond(params: AztecModelParams,
                                     amount=slashed_amount,
                                     kind=TransferKind.slash))
             else:
-                # XXX
+                # NOTE: Costs now include gas fees and safety buffer. 
+                gas: Gas = params['gas_estimators'].commitment_bond(state)
+                fee = gas * state['gas_fee_l1']
+                SAFETY_BUFFER = 2 * fee # HACK:
```
Comment (danlessa) [inline aztec_gddt/logic.py:381]: make that a parameter

File: aztec_gddt/logic.py [hunk 2]
```diff
@@ -372,25 +375,29 @@ def p_commit_bond(params: AztecModelParams,
...
 
                 if payoff_reveal >= 0:
                     # If duration is not expired, do  a trial to see if bond is commited
-                    if bernoulli_trial(probability=params['commit_bond_reveal_probability']) is True and (state['gas_fee_l1'] <= params['gas_threshold_for_tx']):
+                    agent_decides_to_reveal_commit_bond = bernoulli_trial(probability=params['commit_bond_reveal_probability'])
+                    gas_fee_l1_acceptable = (state['gas_fee_l1'] <= params['gas_threshold_for_tx'])
+                    if  agent_decides_to_reveal_commit_bond and gas_fee_l1_acceptable:
                         updated_process = copy(process)
-                        advance_blocks = remaining_time
+                        advance_blocks = remaining_time #TODO: double-check this.
```
Comment (danlessa) [inline aztec_gddt/logic.py:392]: intent: we'll move the L2 time forward by `advance_blocks`

File: aztec_gddt/logic.py [hunk 3]
```diff
@@ -464,45 +473,52 @@ def p_reveal_content(params: AztecModelParams,
...
                                     kind=TransferKind.slash))
             else:
-                # XXX
+                # NOTE: Costs now include gas fees and safety buffer. 
+                gas: Gas = params['gas_estimators'].content_reveal(state)
+                fee = gas * state['gas_fee_l1']
+                SAFETY_BUFFER = 2 * fee # HACK: 
                 expected_rewards = state['cumm_block_rewards'] - history[-1][0]['cumm_block_rewards']
-                expected_costs = state['cumm_fee_cashback'] - history[-1][0]['cumm_fee_cashback']
+                expected_costs = params['op_costs'] + fee + SAFETY_BUFFER
```
Comment (danlessa) [inline aztec_gddt/logic.py:481]: parametrize

File: aztec_gddt/logic.py [hunk 4]
```diff
@@ -464,45 +473,52 @@ def p_reveal_content(params: AztecModelParams,
                                     amount=slashed_amount,
                                     kind=TransferKind.slash))
             else:
-                # XXX
+                # NOTE: Costs now include gas fees and safety buffer. 
+                gas: Gas = params['gas_estimators'].content_reveal(state)
+                fee = gas * state['gas_fee_l1']
+                SAFETY_BUFFER = 2 * fee # HACK: 
                 expected_rewards = state['cumm_block_rewards'] - history[-1][0]['cumm_block_rewards']
```
Comment (danlessa) [inline aztec_gddt/logic.py:480]: add XXX in here: expected rewards = reward on last timestep

File: aztec_gddt/params.py [hunk 5]
```diff
@@ -80,14 +85,130 @@
...
+    else:
+        return single_shock_gas_fee_l1_time_series[-1]
+
+def single_shock_blob_gas_estimate(state: AztecModelState):
+    if state['timestep'] < len(single_shock_gas_fee_blob_time_series):
+        return single_shock_gas_fee_blob_time_series[state['timestep']]
+    else:
+        return single_shock_gas_fee_blob_time_series[-1]
+
+SINGLE_SHOCK_GAS_ESTIMATORS = L1GasEstimators(
```
Comment (danlessa) [inline aztec_gddt/params.py:162]: note: the above estimators are relevant for the Gas Fee, not for the Gas Units per Tx

File: aztec_gddt/params.py [hunk 6]
```diff
@@ -80,14 +85,130 @@
...
+
+
+intermit_shock_gas_fee_l1_time_series = np.zeros(TIMESTEPS)
+intermit_shock_gas_fee_blob_time_series = np.zeros(TIMESTEPS)
+
+intermit_shock_gas_fee_l1_time_series[0:initial_time] = steady_gas_fee_l1_time_series[0:initial_time].copy()
+intermit_shock_gas_fee_l1_time_series[-final_time:] = steady_gas_fee_l1_time_series[-final_time:].copy()
+intermit_shock_gas_fee_l1_time_series[initial_time:TIMESTEPS - final_time] = steady_gas_fee_l1_time_series[initial_time:TIMESTEPS - final_time].copy() + L1_INTER_SHOCK_SIGNAL
+
+def intermit_shock_l1_gas_estimate(state: AztecModelState):
```
Comment (danlessa) [inline aztec_gddt/params.py:194]: note: the above estimators are relevant for the Gas Fee, not for the Gas Units per Tx

File: aztec_gddt/params.py [hunk 7]
```diff
@@ -104,11 +225,16 @@
                                      fee_subsidy_fraction=1.0, # TODO
 
                                      # Phase Durations
-                                     phase_duration_proposal=10, # TODO
-                                     phase_duration_reveal=10, # TODO
-                                     phase_duration_commit_bond=10, # TODO
-                                     phase_duration_rollup=30, # TODO
-                                     phase_duration_race=30, # TODO
+                                     phase_duration_proposal_min_blocks=0, # TODO
```
Comment (danlessa) [inline aztec_gddt/params.py:228]: update numbers to reflect the workplan https://hackmd.io/@blockscience/B1ZQ_TX3T

File: aztec_gddt/types.py [hunk 8]
```diff
@@ -262,17 +263,25 @@ class AztecModelParams(TypedDict):
...
+    phase_duration_rollup_max_blocks: L1Blocks
+    phase_duration_race_min_blocks: L1Blocks
+    phase_duration_race_max_blocks: L1Blocks
 
     stake_activation_period: L1Blocks  # XXX
     unstake_cooldown_period: L1Blocks  # XXX
 
     # Behavioral Parameters
 
+    logic: Dict[str, Callable[Dict, Any]] = None #placeholder for general system logic
```
Comment (danlessa) [inline aztec_gddt/types.py:283]: what is general system logic?

**Review by danlessa — CHANGES_REQUESTED** [review]

**Pull request merged**